File: Java/1414. Find the Minimum Number of Fibonacci Numbers Whose Sum Is K.java
import java.util.Arrays;

class Solution {
    public int findMinFibonacciNumbers(int k) {
        int res = 0;
        List<Integer> fib = new ArrayList<>();
        fib.add(1); fib.add(1);

        while(fib.get(fib.size() - 1) < k) fib.add(fib.get(fib.size() - 1) + fib.get(fib.size() - 2));
        while(k > 0){
            for(int i = fib.size() - 1; i >= 0; i--){
                if(fib.get(i) <= k){
                    k -= fib.get(i);
                    res++;
                }
            }
        }

        return res;
    }
}

File: Python/1414. Find the Minimum Number of Fibonacci Numbers Whose Sum Is K.py
class Solution:
    def findMinFibonacciNumbers(self, k: int) -> int:
        fib, res = [1, 1], 0

        while fib[-1] < k: fib.append(fib[-2] + fib[-1])

        l = len(fib)
        while k > 0:
            for i in range(l - 1, -1, -1):
                if fib[i] <= k:
                    k -= fib[i]
                    res += 1
                    break

        return res
